midend/has_side_effects.h

```diff
@@ -18,18 +18,38 @@ limitations under the License.
 #define MIDEND_HAS_SIDE_EFFECTS_H_
 
 #include "ir/ir.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/common/resolveReferences/referenceMap.h"
 
-/* Should this be a method on IR::Expression? */
+/* Should this be a method on IR::Expression?  Maybe after the refMap/typeMap go away */
 
 class hasSideEffects : public Inspector {
+    P4::ReferenceMap    *refMap = nullptr;
+    P4::TypeMap         *typeMap = nullptr;
+
     bool result = false;
     bool preorder(const IR::AssignmentStatement *) override { return !(result = true); }
```

Comment:
My preference is for `result = true; return false;`. As currently written, this is easily misinterpreted as being equivalent to `return !result;`.